backend-rust/src/indexer.rs

```diff
@@ -0,0 +1,779 @@
+struct PreparedBlock {
+    hash:                 String,
+    height:               i64,
+    slot_time:            NaiveDateTime,
+    baker_id:             Option<i64>,
+    total_amount:         i64,
+    total_staked:         i64,
+    block_last_finalized: String,
+    prepared_block_items: Vec<PreparedBlockItem>,
+}
+
+impl PreparedBlock {
+    fn prepare(data: &BlockData) -> anyhow::Result<Self> {
+        let height = i64::try_from(data.finalized_block_info.height.height)?;
+        let hash = data.finalized_block_info.block_hash.to_string();
+        let block_last_finalized = data.block_info.block_last_finalized.to_string();
+        let slot_time = data.block_info.block_slot_time.naive_utc();
+        let baker_id = if let Some(index) = data.block_info.block_baker {
+            Some(i64::try_from(index.id.index)?)
+        } else {
+            None
+        };
+        let common_reward_data = match data.tokenomics_info {
+            RewardsOverview::V0 {
+                data,
+            } => data,
+            RewardsOverview::V1 {
+                common,
+                ..
+            } => common,
+        };
+        let total_amount = i64::try_from(common_reward_data.total_amount.micro_ccd())?;
+        let total_staked = match data.tokenomics_info {
+            RewardsOverview::V0 {
+                ..
+            } => {
+                // TODO Compute the total staked capital.
+                0i64
+            }
+            RewardsOverview::V1 {
+                total_staked_capital,
+                ..
+            } => i64::try_from(total_staked_capital.micro_ccd())?,
+        };
+
+        let mut prepared_block_items = Vec::new();
+        for block_item in data.events.iter() {
+            prepared_block_items.push(PreparedBlockItem::prepare(data, block_item)?)
+        }
+
+        Ok(Self {
+            hash,
+            height,
+            slot_time,
+            baker_id,
+            total_amount,
+            total_staked,
+            block_last_finalized,
+            prepared_block_items,
+        })
+    }
+
+    async fn save(
+        &self,
+        context: &mut BlockProcessor,
+        tx: &mut sqlx::Transaction<'static, sqlx::Postgres>,
+    ) -> anyhow::Result<()> {
+        sqlx::query!(
+            r#"INSERT INTO blocks (height, hash, slot_time, block_time, baker_id, total_amount, total_staked)
+VALUES ($1, $2, $3,
+  (SELECT EXTRACT("MILLISECONDS" FROM $3 - b.slot_time) FROM blocks b WHERE b.height=($1 - 1::bigint)),
+  $4, $5, $6);"#,
+            self.height,
+            self.hash,
+            self.slot_time,
+            self.baker_id,
+            self.total_amount,
+            self.total_staked
+        )
+        .execute(tx.as_mut())
+            .await?;
+
+        // Check if this block knows of a new finalized block.
+        // If so, mark the blocks since last time as finalized by this block.
+        if self.block_last_finalized != context.last_finalized_hash {
```

Comment:
yes, in the old consensus it was very likely, but it is also possible in the new consensus, since finalization requires two executive blocks (i.e. no timeout certificates in between).
The indexer only processes blocks that are finalized, so no rollback is needed, but we need to compute the finalization time and this is the best that I could come up with.